core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2136,9 +2137,24 @@ private[spark] class DAGScheduler(
     }
   }
 
-  private[scheduler] def handleShuffleMergeFinalized(stage: ShuffleMapStage): Unit = {
-    stage.shuffleDep.markShuffleMergeFinalized
-    processShuffleMapStageCompletion(stage)
+  private[scheduler] def handleRegisterMergeStatuses(
+      stage: ShuffleMapStage,
+      mergeStatuses: Seq[(Int, MergeStatus)]): Unit = {
+    // Register merge statuses if the stage is still running and shuffle merge is not finalized yet.
+    if (runningStages.contains(stage) && !stage.shuffleDep.shuffleMergeFinalized) {
```

Comment:
@mridulm mentioned that one of the benefits of this way is that we can still accept and register merge statuses which are received after the stage gets finalized.
But here, we are ignoring these  statuses if shuffle merge is finalized. Why?
 In case of deterministic stage retries, accepting merge statuses after finalize is still okay, isn't it.

Reply:
Let us do it as a follow up work, given the stage would have completed by the time that comes through.
Can you add a TODO for this @venkata91 ?

Note - the check `runningStages.contains(stage)` is not sufficient - as I detail below for `handleShuffleMergeFinalized`: the same condition applies here too.

Reply:
In that case, currently there is no need to even post these messages to `eventProcessLoop` after `ShuffleMergeFinalized` message is posted. 
So should `RegisterMergeStatuses`  be posted only when `!timedOut.get()`?  Anyways, these messages are just getting ignored so why to even deserialize them.